dvc/remote/gdrive/pydrive.py

```diff
@@ -0,0 +1,93 @@
+import os
+
+from dvc.remote.gdrive.utils import TrackFileReadProgress, FOLDER_MIME_TYPE
+
+
+class RequestBASE:
+    def __init__(self, drive):
+        self.drive = drive
+
+    def execute(self):
+        raise NotImplementedError
+
+
+class RequestListFile(RequestBASE):
+    def __init__(self, drive, query):
+        super(RequestListFile, self).__init__(drive)
+        self.query = query
+
+    def execute(self):
+        return self.drive.ListFile(
+            {"q": self.query, "maxResults": 1000}
+        ).GetList()
+
+
+class RequestListFilePaginated(RequestBASE):
+    def __init__(self, drive, query):
+        super(RequestListFilePaginated, self).__init__(drive)
+        self.query = query
+        self.iter = None
+
+    def execute(self):
+        if not self.iter:
+            self.iter = iter(
+                self.drive.ListFile({"q": self.query, "maxResults": 1000})
+            )
+        return next(self.iter, None)
+
+
+class RequestUploadFile(RequestBASE):
+    def __init__(
+        self, args, no_progress_bar=True, from_file="", progress_name=""
+    ):
+        super(RequestUploadFile, self).__init__(args["drive"])
+        self.title = args["title"]
+        self.parent_id = args["parent_id"]
+        self.mime_type = args["mime_type"]
+        self.no_progress_bar = no_progress_bar
+        self.from_file = from_file
+        self.proress_name = progress_name
+
+    def upload(self, item):
+        with open(self.from_file, "rb") as from_file:
+            if not self.no_progress_bar:
+                from_file = TrackFileReadProgress(self.proress_name, from_file)
+            if os.stat(self.from_file).st_size:
+                item.content = from_file
+            item.Upload()
+
+    def execute(self):
+        item = self.drive.CreateFile(
+            {
+                "title": self.title,
+                "parents": [{"id": self.parent_id}],
+                "mimeType": self.mime_type,
+            }
+        )
+        if self.mime_type == FOLDER_MIME_TYPE:
```

Comment:
could you please elaborate a little bit what does FOLDER_MIME_TYPE mean what is the difference here?

Reply:
GDrive remote item has property `mimeType` which can be one of the following https://developers.google.com/drive/api/v2/mime-types

`application/vnd.google-apps.folder` should be set as `mimeType` if we want to upload new folder to GDrive. In comparison with file creation, creation of folder doesn't require uploading of local file's content.

Reply:
would be probably easier to read if we had this check inside the upload method itself

so, am I correct that in case of folder it's not upload it's more like a folder create ... it's not uploading files inside the folder, only creates it? thus I was confused with a name probably ... would be easier to create a seprate request - RequestCreateFolder instead of RequestUploadFile to handle that specific case ... this way the client of this layer would not need to know anything about some mime types, etc

Reply:
> so, am I correct that in case of folder it's not upload it's more like a folder create ... it's not uploading files inside the folder, only creates it? 

Yes, `Upload` just creates folder or uploads+creates file.

> would be easier to create a seprate request - RequestCreateFolder instead of RequestUploadFile to handle that specific case

Sounds good.